File: apps/android/app/src/main/java/com/avradio/core/database/AVRadioAppDataService.kt
package com.avradio.core.database

import com.avradio.core.access.AVAppsAPIClient
import com.avradio.core.model.Station
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

private object AVRadioAppDataConstants {
    const val appId = "avradio"
    const val resource = "library"
    const val deviceId = "avradio-android"
}

data class AVRadioLibraryDocument(
    val snapshot: AVRadioLibrarySnapshot?,
    val updatedAt: String
)

@Serializable
data class AVRadioLibrarySnapshot(
    val favorites: List<FavoriteStationRecord>,
    val recents: List<RecentStationRecord>,
    val settings: AppSettingsRecord
) {
    val hasMeaningfulContent: Boolean
        get() = favorites.isNotEmpty() || recents.isNotEmpty() || settings.hasMeaningfulContent
}

@Serializable
data class FavoriteStationRecord(
    val station: StationRecord,
    val createdAt: String
)

@Serializable
data class RecentStationRecord(
    val station: StationRecord,
    val lastPlayedAt: String
)

@Serializable
data class AppSettingsRecord(
    val preferredCountry: String,
    val preferredLanguage: String,
    val preferredTag: String,
    val lastPlayedStationID: String?,
    val sleepTimerMinutes: Int?,
    val updatedAt: String
) {
    val hasMeaningfulContent: Boolean
        get() = preferredCountry.isNotEmpty() ||
            preferredLanguage.isNotEmpty() ||
            preferredTag.isNotEmpty() ||
            lastPlayedStationID != null ||
            sleepTimerMinutes != null
}

@Serializable
data class StationRecord(
    val id: String,
    val name: String,
    val country: String,
    val countryCode: String? = null,
    val state: String? = null,
    val language: String,
    val languageCodes: String? = null,
    val tags: String,
    @SerialName("streamURL")
    val streamUrl: String,
    @SerialName("faviconURL")
    val faviconUrl: String? = null,
    val bitrate: Int? = null,
    val codec: String? = null,
    @SerialName("homepageURL")
    val homepageUrl: String? = null,
    val votes: Int? = null,
    val clickCount: Int? = null,
    val clickTrend: Int? = null,
    @SerialName("isHLS")
    val isHls: Boolean? = null,
    val hasExtendedInfo: Boolean? = null,
    @SerialName("hasSSLError")
    val hasSslError: Boolean? = null,
    @SerialName("lastCheckOKAt")
    val lastCheckOkAt: String? = null,
    val geoLatitude: Double? = null,
    val geoLongitude: Double? = null
)

@Serializable
private data class AppDataResponsePayload(
    val data: AppDataEnvelopePayload,
    val updatedAt: String
)

@Serializable
private data class AppDataEnvelopePayload(
    val appId: String,
    val resource: String,
    val deviceId: String,
    val sentAt: String,
    val entries: List<AVRadioLibrarySnapshot>
)

class AVRadioAppDataService(
    private val apiClient: AVAppsAPIClient
) {
    fun isConfigured(): Boolean = apiClient.isConfigured()

    suspend fun pullLibrary(): AVRadioLibraryDocument? {
        val payload = apiClient.request(
            path = "/v1/apps/${AVRadioAppDataConstants.appId}/data/${AVRadioAppDataConstants.resource}",
            serializer = AppDataResponsePayload.serializer()
        ) ?: return null

        return AVRadioLibraryDocument(
            snapshot = payload.data.entries.firstOrNull(),
            updatedAt = payload.updatedAt
        )
    }

    suspend fun pushLibrary(snapshot: AVRadioLibrarySnapshot) {
        val envelope = AppDataEnvelopePayload(
            appId = AVRadioAppDataConstants.appId,
            resource = AVRadioAppDataConstants.resource,
            deviceId = AVRadioAppDataConstants.deviceId,
            sentAt = isoString(nowMillis()),
            entries = listOf(snapshot)
        )

        apiClient.request(
            path = "/v1/apps/${AVRadioAppDataConstants.appId}/data/${AVRadioAppDataConstants.resource}",
            serializer = AppDataResponsePayload.serializer(),
            method = "PUT",
            body = apiClient.encodeToString(AppDataEnvelopePayload.serializer(), envelope)
        )
    }

    companion object {
        fun isoString(epochMillis: Long): String = java.time.Instant.ofEpochMilli(epochMillis).toString()

        fun nowMillis(): Long = System.currentTimeMillis()

        fun epochMillis(value: String): Long =
            runCatching { java.time.Instant.parse(value).toEpochMilli() }.getOrDefault(0L)
    }
}

val Station.appDataRecord: StationRecord
    get() = StationRecord(
        id = id,
        name = name,
        country = country,
        countryCode = countryCode,
        state = state,
        language = language,
        languageCodes = languageCodes,
        tags = tags,
        streamUrl = streamUrl,
        faviconUrl = faviconUrl,
        bitrate = bitrate,
        codec = codec,
        homepageUrl = homepageUrl,
        votes = votes,
        clickCount = clickCount,
        clickTrend = clickTrend,
        isHls = isHls,
        hasExtendedInfo = hasExtendedInfo,
        hasSslError = hasSslError,
        lastCheckOkAt = lastCheckOkAt,
        geoLatitude = geoLatitude,
        geoLongitude = geoLongitude
    )

fun Station(record: StationRecord): Station = Station(
    id = record.id,
    name = record.name,
    country = record.country,
    countryCode = record.countryCode,
    state = record.state,
    language = record.language,
    languageCodes = record.languageCodes,
    tags = record.tags,
    streamUrl = record.streamUrl,
    faviconUrl = record.faviconUrl,
    bitrate = record.bitrate,
    codec = record.codec,
    homepageUrl = record.homepageUrl,
    votes = record.votes,
    clickCount = record.clickCount,
    clickTrend = record.clickTrend,
    isHls = record.isHls,
    hasExtendedInfo = record.hasExtendedInfo,
    hasSslError = record.hasSslError,
    lastCheckOkAt = record.lastCheckOkAt,
    geoLatitude = record.geoLatitude,
    geoLongitude = record.geoLongitude
)

File: apps/ios/Avradio/App/AVRadioAppDataService.swift
import Foundation

private enum AVRadioAppDataConstants {
    static let appId = "avradio"
    static let legacyLibraryResource = "library"
    static let deviceId = "avradio-ios"
}

private enum AVRadioAppDataResource: String, CaseIterable {
    case favorites
    case recents
    case discoveries
    case settings
    case legacyLibrary = "library"

    static let syncResources: [AVRadioAppDataResource] = [
        .favorites,
        .recents,
        .discoveries,
        .settings
    ]
}

struct AVRadioLibraryDocument {
    let snapshot: AVRadioLibrarySnapshot?
    let updatedAt: Date
    let revision: Int
    let etag: String?
}

enum AVRadioLibrarySyncDecision: Equatable {
    case pullRemote(AVRadioLibrarySnapshot)
    case pushLocal
    case noContent
    case alreadyCurrent
}

enum AVRadioLibrarySyncPlanner {
    static func decision(
        localSnapshot: AVRadioLibrarySnapshot,
        localUpdatedAt: Date,
        remoteDocument: AVRadioLibraryDocument
    ) -> AVRadioLibrarySyncDecision {
        let localHasContent = localSnapshot.hasMeaningfulContent

        guard let remoteSnapshot = remoteDocument.snapshot else {
            return localHasContent ? .pushLocal : .noContent
        }

        let remoteHasContent = remoteSnapshot.hasMeaningfulContent
        if !remoteHasContent {
            return localHasContent ? .pushLocal : .noContent
        }

        if !localHasContent || remoteDocument.updatedAt > localUpdatedAt {
            return .pullRemote(remoteSnapshot)
        }

        if localUpdatedAt > remoteDocument.updatedAt {
            return .pushLocal
        }

        return .alreadyCurrent
    }
}

enum AVRadioAppDataError: Error {
    case conflict
}

struct AVRadioLibrarySnapshot: Codable, Equatable {
    let favorites: [FavoriteStationRecord]
    let recents: [RecentStationRecord]
    let discoveries: [DiscoveredTrackRecord]
    let settings: AppSettingsRecord

    var hasMeaningfulContent: Bool {
        !favorites.isEmpty || !recents.isEmpty || !discoveries.isEmpty || settings.hasMeaningfulContent
    }

    init(
        favorites: [FavoriteStationRecord],
        recents: [RecentStationRecord],
        discoveries: [DiscoveredTrackRecord] = [],
        settings: AppSettingsRecord
    ) {
        self.favorites = favorites
        self.recents = recents
        self.discoveries = discoveries
        self.settings = settings
    }

    private enum CodingKeys: String, CodingKey {
        case favorites
        case recents
        case discoveries
        case settings
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        favorites = try container.decode([FavoriteStationRecord].self, forKey: .favorites)
        recents = try container.decode([RecentStationRecord].self, forKey: .recents)
        discoveries = try container.decodeIfPresent([DiscoveredTrackRecord].self, forKey: .discoveries) ?? []
        settings = try container.decode(AppSettingsRecord.self, forKey: .settings)
    }
}

struct FavoriteStationRecord: Codable, Equatable {
    let station: StationRecord
    let createdAt: String
}

struct RecentStationRecord: Codable, Equatable {
    let station: StationRecord
    let lastPlayedAt: String
}

struct DiscoveredTrackRecord: Codable, Equatable {
    let discoveryID: String
    let title: String
    let artist: String?
    let stationID: String
    let stationName: String
    let artworkURL: String?
    let stationArtworkURL: String?
    let playedAt: String
    let markedInterestedAt: String?
    let hiddenAt: String?
}

struct AppSettingsRecord: Codable, Equatable {
    let preferredCountry: String
    let preferredLanguage: String
    let preferredTag: String
    let lastPlayedStationID: String?
    let sleepTimerMinutes: Int?
    let updatedAt: String

    var hasMeaningfulContent: Bool {
        !preferredCountry.isEmpty ||
            !preferredLanguage.isEmpty ||
            !preferredTag.isEmpty ||
            lastPlayedStationID != nil ||
            sleepTimerMinutes != nil
    }

    static var empty: AppSettingsRecord {
        AppSettingsRecord(
            preferredCountry: "",
            preferredLanguage: "",
            preferredTag: "",
            lastPlayedStationID: nil,
            sleepTimerMinutes: nil,
            updatedAt: "1970-01-01T00:00:00.000Z"
        )
    }
}

struct StationRecord: Codable, Equatable {
    let id: String
    let name: String
    let country: String
    let countryCode: String?
    let state: String?
    let language: String
    let languageCodes: String?
    let tags: String
    let streamURL: String
    let faviconURL: String?
    let bitrate: Int?
    let codec: String?
    let homepageURL: String?
    let votes: Int?
    let clickCount: Int?
    let clickTrend: Int?
    let isHLS: Bool?
    let hasExtendedInfo: Bool?
    let hasSSLError: Bool?
    let lastCheckOKAt: String?
    let geoLatitude: Double?
    let geoLongitude: Double?
}

private struct AppDataResponsePayload<Entry: Codable>: Decodable {
    let data: AppDataEnvelopePayload<Entry>
    let updatedAt: String
    let revision: Int?
    let etag: String?
}

private struct AppDataEnvelopePayload<Entry: Codable>: Codable {
    let appId: String
    let resource: String
    let deviceId: String
    let sentAt: String
    let entries: [Entry]
}

private struct AppDataResourceDocument<Entry: Codable> {
    let entries: [Entry]
    let updatedAt: Date
    let revision: Int
    let etag: String?
}

@MainActor
final class AVRadioAppDataService {
    private let apiClient: AVAppsAPIClient
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private var lastKnownRevisions: [String: Int] = [:]
    private var lastKnownEtags: [String: String] = [:]

    init(apiClient: AVAppsAPIClient) {
        self.apiClient = apiClient
    }

    func isConfigured() -> Bool {
        apiClient.isConfigured()
    }

    func pullLibrary() async throws -> AVRadioLibraryDocument {
        let favorites = try await pullResource(
            .favorites,
            entryType: FavoriteStationRecord.self
        )
        let recents = try await pullResource(
            .recents,
            entryType: RecentStationRecord.self
        )
        let discoveries = try await pullResource(
            .discoveries,
            entryType: DiscoveredTrackRecord.self
        )
        let settings = try await pullResource(
            .settings,
            entryType: AppSettingsRecord.self
        )

        if [favorites.revision, recents.revision, discoveries.revision, settings.revision].allSatisfy({ $0 == 0 }) {
            return try await pullLegacyLibrary()
        }

        let snapshot = AVRadioLibrarySnapshot(
            favorites: favorites.entries,
            recents: recents.entries,
            discoveries: discoveries.entries,
            settings: settings.entries.first ?? .empty
        )
        let updatedAt = [
            favorites.updatedAt,
            recents.updatedAt,
            discoveries.updatedAt,
            settings.updatedAt
        ].max() ?? .distantPast

        return AVRadioLibraryDocument(
            snapshot: snapshot.hasMeaningfulContent ? snapshot : nil,
            updatedAt: updatedAt,
            revision: [
                favorites.revision,
                recents.revision,
                discoveries.revision,
                settings.revision
            ].max() ?? 0,
            etag: nil
        )
    }

    func pushLibrary(_ snapshot: AVRadioLibrarySnapshot) async throws {
        try await pushResource(.favorites, entries: snapshot.favorites)
        try await pushResource(.recents, entries: snapshot.recents)
        try await pushResource(.discoveries, entries: snapshot.discoveries)
        try await pushResource(.settings, entries: [snapshot.settings])
    }

    func overwriteLibrary(_ snapshot: AVRadioLibrarySnapshot) async throws {
        for resource in AVRadioAppDataResource.syncResources {
            forgetSyncVersion(for: resource)
        }

        try await pushLibrary(snapshot)
    }

    private func pullLegacyLibrary() async throws -> AVRadioLibraryDocument {
        let payload: AppDataResponsePayload<AVRadioLibrarySnapshot> = try await apiClient.request(
            path: dataPath(for: .legacyLibrary)
        )
        rememberSyncVersion(
            for: .legacyLibrary,
            revision: payload.revision,
            etag: payload.etag
        )

        return AVRadioLibraryDocument(
            snapshot: payload.data.entries.first,
            updatedAt: Self.date(from: payload.updatedAt),
            revision: payload.revision ?? 0,
            etag: payload.etag
        )
    }

    private func pullResource<Entry: Codable>(
        _ resource: AVRadioAppDataResource,
        entryType: Entry.Type
    ) async throws -> AppDataResourceDocument<Entry> {
        let payload: AppDataResponsePayload<Entry> = try await apiClient.request(
            path: dataPath(for: resource)
        )
        rememberSyncVersion(
            for: resource,
            revision: payload.revision,
            etag: payload.etag
        )

        return AppDataResourceDocument(
            entries: payload.data.entries,
            updatedAt: Self.date(from: payload.updatedAt),
            revision: payload.revision ?? 0,
            etag: payload.etag
        )
    }

    private func pushResource<Entry: Codable>(
        _ resource: AVRadioAppDataResource,
        entries: [Entry]
    ) async throws {
        let envelope = AppDataEnvelopePayload(
            appId: AVRadioAppDataConstants.appId,
            resource: resource.rawValue,
            deviceId: AVRadioAppDataConstants.deviceId,
            sentAt: Self.isoString(from: .now),
            entries: entries
        )

        var headers: [String: String] = [:]
        if let lastKnownEtag = lastKnownEtags[resource.rawValue] {
            headers["If-Match"] = lastKnownEtag
        } else if let lastKnownRevision = lastKnownRevisions[resource.rawValue] {
            headers["If-Match"] = "\"revision-\(lastKnownRevision)\""
        }

        let response: AppDataResponsePayload<Entry>
        do {
            response = try await apiClient.request(
                path: dataPath(for: resource),
                method: "PUT",
                body: try encoder.encode(envelope),
                headers: headers
            )
        } catch AVAppsAPIClientError.requestFailed(let statusCode) where statusCode == 409 {
            throw AVRadioAppDataError.conflict
        }
        rememberSyncVersion(for: resource, revision: response.revision, etag: response.etag)
    }

    private func rememberSyncVersion(
        for resource: AVRadioAppDataResource,
        revision: Int?,
        etag: String?
    ) {
        lastKnownRevisions[resource.rawValue] = revision
        lastKnownEtags[resource.rawValue] = etag
    }

    private func forgetSyncVersion(for resource: AVRadioAppDataResource) {
        lastKnownRevisions[resource.rawValue] = nil
        lastKnownEtags[resource.rawValue] = nil
    }

    private func dataPath(for resource: AVRadioAppDataResource) -> String {
        "/v1/apps/\(AVRadioAppDataConstants.appId)/data/\(resource.rawValue)"
    }

    private static func date(from value: String) -> Date {
        AVRadioDateCoding.date(from: value)
    }

    static func isoString(from date: Date) -> String {
        AVRadioDateCoding.string(from: date)
    }
}
